pkg/webhooks/imagecontentpolicies/imagecontentpolicies.go

```diff
@@ -0,0 +1,188 @@
+	admissionregv1 "k8s.io/api/admissionregistration/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+const (
+	WebhookName = "imagecontentpolicies-validation"
+	WebhookDoc  = "Managed OpenShift customers may not create ImageContentPolicy or ImageContentSourcePolicy resources that configure mirrors for quay.io, registry.redhat.com, nor registry.access.redhat.com."
+	// unauthorizedRepositoryMirrors is a regex that is used to reject certain specified repository mirrors.
+	// Generally all contained regexes follow a similar pattern, i.e. rejecting quay.io or quay.io/.*
+	unauthorizedRepositoryMirrors = `(^quay\.io(/.*)?$|^registry\.redhat\.io(/.*)?$|^registry.access.redhat.com(/.*)?)`
+)
+
+type ImageContentPoliciesWebhook struct {
+	scheme *runtime.Scheme
+	log    logr.Logger
+}
+
+func NewWebhook() *ImageContentPoliciesWebhook {
+	return &ImageContentPoliciesWebhook{
+		scheme: runtime.NewScheme(),
+		log:    logf.Log.WithName(WebhookName),
+	}
+}
+
+func (w *ImageContentPoliciesWebhook) Authorized(request admission.Request) admission.Response {
+	decoder, err := admission.NewDecoder(w.scheme)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	icp := configv1.ImageContentPolicy{}
+	icsp := operatorv1alpha1.ImageContentSourcePolicy{}
+	allowedResp := admission.Response{
+		AdmissionResponse: admissionv1.AdmissionResponse{
+			UID:     request.UID,
+			Allowed: true,
+		},
+	}
+
+	if len(request.Object.Raw) > 0 {
+		switch request.RequestKind.Kind {
+		case "ImageContentPolicy":
+			if err := decoder.Decode(request, &icp); err != nil {
+				w.log.Error(err, "failed to render an ImageContentPolicy from request")
+				return admission.Errored(http.StatusBadRequest, err)
+			}
+
+			if !authorizeImageContentPolicy(icp) {
+				return admission.Errored(http.StatusBadRequest, fmt.Errorf(WebhookDoc))
+			}
+		case "ImageContentSourcePolicy":
+			if err := decoder.Decode(request, &icsp); err != nil {
+				w.log.Error(err, "failed to render an ImageContentSourcePolicy from request")
+				return admission.Errored(http.StatusBadRequest, err)
+			}
+
+			if !authorizeImageContentSourcePolicy(icsp) {
+				return admission.Errored(http.StatusBadRequest, fmt.Errorf(WebhookDoc))
+			}
+		}
+	}
+
+	return allowedResp
+}
+
+func (w *ImageContentPoliciesWebhook) GetURI() string {
+	return "/" + WebhookName
+}
+
+func (w *ImageContentPoliciesWebhook) Validate(request admission.Request) bool {
+	switch {
+	case request.Kind.Kind == "ImageContentPolicy":
+	case request.Kind.Kind == "ImageContentSourcePolicy":
+		return true
+	default:
+		return false
+	}
+
+	return false
+}
+
+func (w *ImageContentPoliciesWebhook) Name() string {
+	return WebhookName
+}
+
+func (w *ImageContentPoliciesWebhook) FailurePolicy() admissionregv1.FailurePolicyType {
+	// Fail-closed
+	return admissionregv1.Fail
```

Comment:
Most other webhooks I've seen in this repo are `admissionregv1.Ignore` i.e. "fail-open".

There are definitely cons to fail-closed if this webhook is failing, however is it better if we fail-open and a problematic`ImageContentPolicy` gets created is that preferred? I was leaning towards fail-closed to start out with and we can move it to fail-open if UX concerns arise and as we add more observability into how these CRs interact with existing alerts?